File: MATLAB/Shift_German.m
% Shift cipher for the German alphabet
% Nightsong by Friedrich Nietzsche
fd=fopen('nachtlied-utf-8.txt','r');
x=fread(fd);
fclose(fd);
y=PreProcess_German(x');
k=23;

% A true ASCII shift is performed as follows:
%z= y-97+k +65;
%y1 = z-65-k+97;
%cat(1, char(y(300:400)), char(z(300:400)), char(y1(300:400)))

% An alphabet shift is harder because we need to grab the extended ASCII
% characters and convert them so the modulus operation works properly.  We
% will perform substitution to constrain them to {1...30} instead 
% of {1...155}

y2 = ASCII_30_Add(y);
%z1 = mod(y2-97+k,30)+65;
z1 = EncryptShift_German(y2, k);
z2 = ASCII_Enc_German(z1);
z3 = ASCII_Dec_German(z2);
%y3 = mod(z1-65-k,30)+97;
y3 = DecryptShift_German(z1, k);
y4 = ASCII_30_Del(y3);

cat(1, char(y(300:400)),  char(z2(300:400)),  char(y4(300:400)))

File: JavaScript/Shift_German.js
// Import required functions
import { PreProcess_German } from "./PreProcess_German.js"
import { ASCII_30_Add } from "./ASCII_30_Add.js"
import { EncryptShift_German } from "./EncryptShift_German.js"
import { ASCII_Enc_German } from "./ASCII_Enc_German.js"
import { ASCII_Dec_German } from "./ASCII_Dec_German.js"
import { DecryptShift_German } from "./DecryptShift_German.js"
import { ASCII_30_Del } from "./ASCII_30_Del.js"

// Process Plaintext
var rawtext = fs.readFileSync("nachtlied-utf-8.txt", "utf-8");
var plaintext = PreProcess_German(rawtext);
var plaintext_ascii = plaintext.split('').map(function (char) { return char.charCodeAt(0); });
var plaintext_ascii_30 = plaintext_ascii.map(ASCII_30_Add);

// Encrypt Plaintext
var ciphertext_ascii_30 = plaintext_ascii_30.map(EncryptShift_German);
var ciphertext_ascii = ciphertext_ascii_30.map(ASCII_Enc_German);
var ciphertext = String.fromCharCode.apply(null, ciphertext_ascii);

// Decrypt Ciphertext
ciphertext_ascii_30 = ciphertext_ascii.map(ASCII_Dec_German);
var dectext_ascii_30 = ciphertext_ascii_30.map(DecryptShift_German);
var dectext_ascii = dectext_ascii_30.map(ASCII_30_Del);
var dectext = String.fromCharCode.apply(null, dectext_ascii);

// Decrypt Ciphertext
console.log(dectext);
